[쇼핑 주문 1, 2 단계] 토바에 미션 제출합니다.

tobae-time opened the pull request:
안녕하세요, 페로로!
레벨 1 블랙잭에 이어서 레벨 2에서 다시 만나 뵙게 되어서 영광입니다. 🙇‍♀️
부족하지만 많이 배워가도록 하겠습니다...!

시간이 부족하여 미완성으로 제출하게 되었습니다. 😭
다음 리뷰 요청 때, 구현하지 못한 부분과 테스트 코드도 함께 반영하여 제출하도록 하겠습니다. 
현재 테스트 코드 깨지는 곳이 많습니다. 구현을 우선시 하다 보니, 테스트 코드를 챙기지 못했습니다. 죄송합니다 🥲

2단계의 프로그래밍 요구사항 중 "서버 통신을 위한 JSON 직렬화 라이브러리를 선택하고 PR에 선택 이유를 남긴다."라는 부분이 있습니다.

저희는 Kotlinx.serialization을 선택했습니다. Jetbrains에서 만든 라이브러리이며, 코틀린의 확장 라이브러리라는 점에서 신뢰가 높다고 판단하였습니다. 또한 gson은 reflection 기반으로 런타임에 작동하지만, Kotlinx.serialization는 컴파일 타임에 serialization 코드 생성하기 때문에 속도가 빠르고 메모리 사용이 적습니다. Kotlinx.serialization는 java의 reflection에 의존하지 않아서 jvm 말고도 kotlin multiplatform에서 사용이 가능하기 때문에 선택했습니다.

페어의 코드를 기반으로 해당 미션을 진행했습니다.
[쇼핑 카트 미션 마이그레이션 코드를 제외한 커밋들의 Files Changed](https://github.com/woowacourse/android-shopping-order/pull/104/files/0268b70f516180f51bc6735c1152a610b91560e9..d448684392f786e377807defab5d09e618ce413a)링크 함께 첨부합니다! 리뷰 잘 부탁드립니다! 🙇‍♀️

[쇼핑 주문 1단계 커밋](https://github.com/woowacourse/android-shopping-order/pull/104/files/0268b70f516180f51bc6735c1152a610b91560e9..990160b2ea960d5a668e51baaf9cfef7841a3c48)
[쇼핑 주문 2단계 커밋](https://github.com/woowacourse/android-shopping-order/pull/104/files/990160b2ea960d5a668e51baaf9cfef7841a3c48..d448684392f786e377807defab5d09e618ce413a)

# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?


## 스크린샷
https://github.com/user-attachments/assets/5105f71c-000f-469f-a3e9-3d9f0bea0cfc

chanho0908 committed:
git squash complete

Commit:
chore: 불필요한 코드 제거

chanho0908 committed:
build: add shimmer dependency

chanho0908 committed:
docs: 스켈레톤 UI 노출 기능 요구 사항 추가

chanho0908 committed:
feat: shimmer placeholder UI 추가

- shimmer_background 색상 추가
- shimmer 효과를 위한 item_shimmer_product.xml 레이아웃 추가

chanho0908 committed:
feat: 메인 화면 로딩 중 스켈레톤 UI 표시

메인 화면에서 데이터를 불러오는 동안 스켈레톤 UI를 표시하도록 기능을 추가했습니다.
- `MainViewModel`에 로딩 상태를 관리하는 `_isLoading` LiveData를 추가했습니다.
- `loadInitial` 함수 호출 시 로딩 상태를 true로 변경하고, 데이터 로드 완료 후 false로 변경합니다.
- `MainActivity`에서 `isLoading` LiveData를 관찰하여 로딩 상태에 따라 스켈레톤 UI와 상품 목록 RecyclerView의 가시성을 조절합니다.
- `activity_main.xml`에 ShimmerFrameLayout을 활용한 스켈레톤 UI를 추가했습니다.

Commit:
docs: 서버 연동 기능 요구 사항 추가

Commit:
build: Retrofit, kotlinx serialization 의존성 추가

Commit:
fix: 호환성 버전 해결

Commit:
feat: 상품 목록 네트워크 응답 모델 추가

Commit:
feat: 서버 상품 목록 불러오기 기능 추가

chanho0908 committed:
build: add logging interceptor dependency

chanho0908 committed:
feat: add http logging interceptor

chanho0908 committed:
refactor: Simplify product loading logic in MainViewModel

chanho0908 committed:
refactor: Remove unused ProductWithCartLoader from MainViewModelFactory

chanho0908 committed:
refactor(MainViewModel): 상품 목록 호출 로직 코드 최적화

chanho0908 committed:
feat: BasicAuthentication Interceptor 추가

chanho0908 committed:
feat: 장바구니에 상품을 추가하는 기능 구현

chanho0908 committed:
refactor: 메인 화면 상품 추가 시 서버 연동 로직 추가

- `ProductState`에 `increaseCartQuantity2` 함수를 추가하여 장바구니 수량 증가 시 서버와 동기화하도록 변경했습니다.
- `MainViewModel`에서 상품 추가 시 `cartRepository.addCart`를 호출하여 서버에 장바구니 정보를 전달합니다.
- `MainViewModelFactory`에서 `DefaultCartRepository`를 주입받도록 수정했습니다.
- `MainActivity`에서 `ViewModelProvider.Factory`에 `DefaultCartRepository`를 전달하도록 변경했습니다.
- `ProductUiState`의 `canIncreaseCartQuantity` 함수를 `increaseCartQuantity`로 변경하고, 반환 타입을 `ProductState`로 수정했습니다.

chanho0908 committed:
refactor: AppContainer에 새로운 CartRepository 의존성 추가

chanho0908 committed:
refactor: data 계층의 network 모델을 entity에서 response로 변경

chanho0908 committed:
feat: 장바구니 목록 조회 응답 DTO 추가

장바구니 목록 조회 API 응답을 위한 데이터 클래스들을 추가합니다.
- `CartsResponse`: 페이지네이션 정보를 포함한 전체 응답
- `Content`: 장바구니 상품 정보
- `Product`: 상품 상세 정보
- `Pageable`: 페이지네이션 관련 정보
- `SortX`: 정렬 관련 정보

Commit:
feat: `ProductState`에 `cartId` 추가

Commit:
fix: API 응답 DTO에 Serializable 어노테이션 추가

Commit:
feat: 장바구니 상품 목록 조회 API 응답 모델 정의

Commit:
feat: 장바구니 목록 조회 API 연동

Commit:
refactor: 상품 목록 로드 시 장바구니 정보 병합

상품 목록을 불러올 때 장바구니에 담긴 상품의 수량 정보를 함께 가져와 UI에 반영하도록 수정했습니다.

- `loadNextPage` 함수에서 상품 정보 로드 후 `loadCartAndMerge` 함수를 호출합니다.
- `loadCartAndMerge` 함수는 장바구니 정보를 로드하여 `mergeProductWithCart` 함수를 통해 상품 정보와 병합합니다.
- `mergeProductWithCart` 함수는 각 상품에 대해 장바구니 정보를 찾아 `ProductState`에 반영합니다.
- 로딩 상태 관리 및 오류 로깅 방식을 개선했습니다.

chanho0908 committed:
feat: 장바구니 상품 수량 수정 API 연동

chanho0908 committed:
refactor: 상품 및 장바구니 정보 로드 로직 개선

상품 목록과 장바구니 목록을 동시에 로드하고 병합하도록 로직을 변경했습니다.

- `loadNextPage` 함수를 `loadProductsAndCarts`로 변경하여 상품과 장바구니 정보를 함께 처리합니다.
- `loadCartAndMerge` 함수를 `mergeProductsWithCarts`로 변경하고, 상품과 장바구니 정보를 병합하여 `ProductState` 목록을 생성합니다.
- 로딩 상태 관리 및 오류 처리 방식을 개선하여 `setLoading` 및 `handleError` 함수를 사용합니다.
- 장바구니 수량 증가 시, 이미 장바구니에 담긴 상품이면 수량 업데이트 API를, 아니면 추가 API를 호출하도록 수정했습니다.

chanho0908 committed:
feat: 장바구니 상품 삭제 기능 구현

장바구니에 담긴 특정 상품을 삭제하는 기능을 추가합니다.

chanho0908 committed:
feat: 상품 목록 화면 장바구니 삭제 기능 추가

chanho0908 committed:
refactor: 수량 변경 컴포넌트 데이터 바인딩 방식 변경

수량 변경 컴포넌트(`item_quantity_changer.xml`)에서 `ProductState` 모델 전체를 받는 대신, 필요한 데이터(`cartId`, `quantity`, `containerVisibility`)만 각각 받도록 수정했습니다.

- `item_quantity_changer.xml`:
    - `model` 변수를 제거하고 `cartId`, `quantity`, `containerVisibility` 변수를 추가했습니다.
    - 이벤트 핸들러 및 텍스트, 가시성 설정 시 새로운 변수를 사용하도록 수정했습니다.
- `item_cart.xml`, `item_product.xml`:
    - `item_quantity_changer` 레이아웃을 include 할 때, `bind:model` 대신 새로운 변수들을 개별적으로 바인딩하도록 수정했습니다.
    - `item_cart.xml`: 모델 타입을 `ProductState`에서 `CartState`로 변경하고, `model`의 필드를 직접 참조하도록 수정했습니다.
- `activity_detail.xml`:
    - `item_quantity_changer`를 include 하는 부분을 주석 처리했습니다. (추후 기능 구현 시 재활성화 예정)

chanho0908 committed:
feat: `CartState` 및 `ShoppingCart`를 `CartState`로 변환하는 확장 함수 추가

chanho0908 committed:
feat: 장바구니 화면 상품 추가,삭제, 페이징 기능 API 연동 완료

Commit:
fix: 장바구니에 담긴 상품 수량 감소 시 cartId가 null로 초기화되지 않는 버그 수정

Commit:
feat: 상세 화면 상품 정보 API 연동

chanho0908 committed:
refactor: `ProductState` 및 관련 도메인 모델 수정

- `IncreaseState` 클래스를 삭제하고, 수량 증가 가능 여부 판단 로직을 `ProductState` 내부 또는 사용하는 쪽으로 이동시킵니다.
- `Quantity` 클래스에서 `isExceeded` 함수를 제거합니다. 수량 초과 여부 판단은 해당 로직이 필요한 곳에서 직접 처리하도록 변경합니다.
- `Product` 도메인 클래스에서 `canIncrease` 함수를 제거합니다. `IncreaseState` 삭제에 따른 변경입니다.
- `ProductsDataSource2`의 불필요한 주석을 제거합니다.

chanho0908 committed:
refactor: 상품 상세 화면 수량 변경 로직 수정 및 ViewModel 의존성 정리

- `DetailViewModelFactory` 및 `DetailViewModel`에서 불필요한 `productRepository` 및 `cartRepository` 의존성을 제거했습니다.
- `ProductState`에서 `increaseCartQuantity` 함수의 분기를 제거하고 단순 증가 로직으로 변경했습니다.
- `DetailUiState`에 `cartQuantity`와 `productId` 프로퍼티를 추가하여 데이터 접근을 용이하게 했습니다.
- 상품 상세 화면(`activity_detail.xml`)에서 수량 변경 UI가 ViewModel의 상태를 올바르게 반영하도록 수정했습니다.
- `DetailViewModel`의 `saveCart` 로직을 `DefaultCartRepository`를 사용하도록 변경했습니다.
- `ProductUiState`의 `increaseCartQuantity` 함수 호출을 새로운 `increaseCartQuantity`로 변경했습니다.

chanho0908 committed:
feat: 장바구니 상품 추가 시 ID 반환

장바구니에 상품을 추가하는 `addCart` 함수의 콜백에서 응답 헤더의 Location 값으로부터 cartId를 추출하여 반환하도록 수정했습니다.

chanho0908 committed:
feat: 장바구니 목록 쉬머 아이템 레이아웃 추가

장바구니 목록 로딩 중에 표시될 쉬머 효과를 위한 `item_shimmer_cart.xml` 레이아웃 파일을 추가합니다.

- 상품명과 상품 이미지를 위한 `View` 요소를 포함합니다.
- 배경색과 테두리는 기존 장바구니 아이템과 유사하게 설정합니다.
- 각 `View`의 배경색은 쉬머 효과를 위해 `@color/shimmer_background`로 설정합니다.

chanho0908 committed:
refactor: HistoryLoader 로직 변경

- HistoryLoader에서 `ProductRepository` 대신 `DefaultProductRepository`를 사용하도록 변경했습니다.
- 여러 상품 ID에 대해 `loadProduct`를 호출하고 결과를 취합하여 `HistoryState` 목록을 생성합니다.
- 각 상품 로드 시 발생할 수 있는 오류를 처리하고, 모든 상품 로드가 완료되거나 오류 발생 시 콜백을 호출합니다.
- 콜백은 `Result` 타입으로 변경되어 성공 또는 실패 결과를 전달합니다.

chanho0908 committed:
refactor: CartRepositoryImpl 스레드 제거 및 getCartSinglePage 파라미터 타입 변경

- `CartRepositoryImpl`에서 `thread`를 사용하여 비동기 처리하던 로직을 제거했습니다.
- `CartService`의 `getCartSinglePage` 메서드와 `CartDataSource2`의 `singlePage` 메서드의 `page`, `size` 파라미터 타입을 `Int`에서 `Int?`로 변경하여 선택적 값으로 수정했습니다.

chanho0908 committed:
refactor: `DefaultCartRepository.loadSinglePage` 매개변수 nullable로 변경

`loadSinglePage` 함수의 `page`와 `pageSize` 매개변수를 nullable 타입으로 변경하여, 기본값을 사용하거나 특정 페이지 정보를 전달하지 않을 수 있도록 유연성을 높였습니다.

chanho0908 committed:
feat: 상품 상세 화면에서 장바구니 담기 기능 수정

- 이미 장바구니에 담긴 상품을 또 담을 경우, 수량을 합산하여 업데이트하도록 변경
- 새로운 상품을 담을 경우, 기존 로직과 동일하게 장바구니에 추가

chanho0908 committed:
refactor: 코드 스타일 개선 및 불필요한 코드 제거

chanho0908 committed:
refactor: `ProductState` 및 `ShoppingCart`에 `productId` 접근자 추가

chanho0908 committed:
refactor: 장바구니 수량 동기화 로직 추가

chanho0908 committed:
refactor: DataSource 및 Repository 구조 변경

chanho0908 committed:
refactor: API 인증 정보 및 기본 URL을 `local.properties`에서 로드하도록 변경

chanho0908 committed:
docs: README 기능 요구사항 추가

- 장바구니에서 특정 상품만 골라 주문하는 기능 추가
- 상품 추천 기능 관련 요구사항 상세 명세

chanho0908 committed:
feat: 장바구니 상품 체크박스 기능 추가

장바구니 아이템 레이아웃에 체크박스를 추가하고, 체크 상태 변경 시 `onCheckedChanged` 핸들러를 호출하도록 구현했습니다.

- `item_cart.xml`: 체크박스 UI 추가 및 `onCheckedChanged` 이벤트 핸들러 바인딩
- `CartViewHolder.kt`: `onCheckedChanged` 핸들러 인터페이스 정의

chanho0908 committed:
feat: 장바구니 상품 선택 기능 추가

chanho0908 committed:
feat: 장바구니 화면 하단 주문 기능 및 Fragment 분리

- `CartActivity`의 화면 구성을 `FragmentContainerView`를 사용하도록 변경하여 `CartListFragment`를 포함시켰습니다.
- `CartListFragment`를 추가하여 장바구니 목록과 페이징 UI를 담당하도록 했습니다.
- `CartActivity` 하단에 전체 선택 체크박스, 총 주문 금액, 주문하기 버튼을 포함하는 레이아웃을 추가했습니다.
- `item_cart.xml`에서 수량 변경 UI의 레이아웃 제약을 수정했습니다.
- `strings.xml`에 주문하기 버튼 텍스트(`text_order`)를 추가했습니다.

chanho0908 committed:
fix: 수량 변경 UI 너비 수정

수량 변경 UI의 너비를 `wrap_content`에서 `match_parent`로 변경하여 부모 레이아웃의 너비를 채우도록 수정합니다.

chanho0908 committed:
refactor: CartUiState에 totalPrice, checkedProductCount, setAllItemsChecked 추가

- `totalPrice`: 선택된 상품들의 총 가격을 계산하는 프로퍼티 추가
- `checkedProductCount`: 선택된 상품들의 총 수량을 계산하는 프로퍼티 추가
- `setAllItemsChecked`: 모든 상품의 선택 상태를 한 번에 변경하는 함수 추가

chanho0908 committed:
feat: 주문 완료 화면 레이아웃 추가

- `fragment_order_complete.xml`: 주문 완료 화면 레이아웃 파일 추가
  - 추천 상품 제목, 설명, RecyclerView 포함
- `strings.xml`: 주문 완료 화면에서 사용될 문자열 추가
  - `text_order_complete_recommend_title`: "이런 상품은 어떠세요?"
  - `text_order_complete_recommend_description`: "* 최근 본 상품 기반으로 좋아하실 것 같은 상품들을 추천해드려요."

chanho0908 committed:
feat: 장바구니 상품 페이징 기능 추가 및 전체 선택 상태 유지

- `CartAdapter`가 페이지 번호를 받아 해당 페이지의 아이템만 표시하도록 변경
- `CartViewModel`에 `historyLoader` 추가 및 `loadCarts()` 호출 시점 변경
- `CartViewModel`에서 장바구니 목록 로드 시 전체 선택 상태(`allChecked`)를 `CartState`에 반영
- `CartUiState`에 `allChecked` 프로퍼티 추가 및 `setAllItemsChecked` 호출 시 `allChecked` 상태도 함께 변경
- `CartListFragment`에서 `CartAdapter`에 페이지 번호를 전달하여 페이징된 아이템 목록을 표시하도록 수정

chanho0908 committed:
feat: 주문 완료 화면 추가 및 화면 전환 기능 구현

laco-dev commented on gradle.properties:
```diff
@@ -21,3 +21,4 @@ kotlin.code.style=official
 # resources declared in the library itself and none from the library's dependencies,
 # thereby reducing the size of the R class for that library
 android.nonTransitiveRClass=true
+android.defaults.buildFeatures.buildconfig=true
```
한때 buildconfig가 기본값이 false로 변경되었어요.
왜 그랬을까요? 모든 모듈이 이 값을 사용하지 않기를 바래서였을거에요.
하지만 지금 해결한 코드는 모든 설정이 기본적으로 true가 되도록 한 것이라면 기본값이 바뀐 의도를 잘못 해석했다고 볼 수 있어요.
buildConfig를 사용할 수 있는 다른 설정 방법을 찾아보세요.

laco-dev commented on app/src/main/res/layout/item_shimmer_product.xml:
```diff
@@ -0,0 +1,49 @@
+            android:background="@color/shimmer_background"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
```
그냥 View를 사용하면 될 텐데, Shimmer를 사용하는데 있어 ImageView, TextView가 필요한가요?

tobae-time replied:
형태를 보여주기 위함이므로, View만 사용해도 충분합니다. 이때 작성할 당시에는 item_product.xml을 가져와 변경하여 사용하였습니다. 그 과정에서 고려하지 못했던 것 같습니다. 🥲
[feat: 쉬머 로딩 UI 개선](https://github.com/woowacourse/android-shopping-order/pull/104/commits/7e2dc81a6afeefc019e6e8da3e1d8bd8c0210c3e)

laco-dev commented on app/src/main/res/layout/item_shimmer_cart.xml:
```diff
@@ -0,0 +1,36 @@
+    android:layout_marginHorizontal="18dp"
+    android:background="@drawable/cart_border">
+
+        <View
+            android:id="@+id/text_view_product_name"
+            android:layout_width="220dp"
+            android:layout_height="25dp"
+            android:layout_marginStart="12dp"
+            android:layout_marginTop="18dp"
+            android:text="@{model.name}"
+            android:textColor="@color/gray"
+            android:background="@color/shimmer_background"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            tools:text="[든든] 동원 스위트콘" />
```
반대로 여기서는 View가 쓰이네요..?

tobae-time replied:
shimmer_product를 만든 후 추후에 shimmer_cart를 만들었습니다. shimmer_cart를 만들 당시에는 View여도 충분하다고 판단한 것 같습니다. shimmer_product를 확인하지 않고 진행한 것 같습니다. 😥 앞으로 일관성 챙겨갈 수 있도록 점검하도록 하겠습니다...!

laco-dev commented on app/src/main/res/layout/item_quantity_changer.xml:
```diff
@@ -4,27 +4,37 @@
         android:layout_height="wrap_content"
         android:background="@drawable/product_quantity_background"
-        android:visibility="@{model.hasCartQuantity ? View.VISIBLE : View.GONE}">
+        android:visibility="@{containerVisibility ? View.VISIBLE : View.GONE}">
```
이 아이템을 사용하는 곳에서 visibility를 결정하는 것과 어떤 차이가 있는거에요?

```kotlin
<include
  layout=item_quantity
  visibility=...
```

tobae-time replied:
해당 xml을 사용하는 곳에서 가시성을 변경해주는 게 좋습니다. 해당 뷰를 재사용하는 곳이 상품 목록의 아이템, 상세 페이지, 장바구니의 아이템, 추천 상품의 아이템 4곳입니다. 이 중 가시성 제어가 필요한 곳은 상품 목록의 아이템, 추천 상품의 아이템 2곳입니다. 뷰의 일부 가시성을 제어하는 것이 아닌 item_quantity_changer.xml 전체의 가시성을 제어해야 하므로 사용하는 곳에서 visibility를 결정하는 것이 적절한 선택이라고 생각됩니다!
[refactor: item_quantity_changer.xml을 사용하는 곳에서 가시성을 제어하도록 변경](https://github.com/woowacourse/android-shopping-order/pull/104/commits/7b9bea046bedf6151203e2ca3ea625530831e2dc)

laco-dev commented on app/src/main/res/layout/fragment_cart_list.xml:
```diff
@@ -0,0 +1,105 @@
+            app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
+            app:layout_constraintBottom_toTopOf="@id/pager"
+            app:layout_constraintTop_toTopOf="parent" />
+
+        <com.facebook.shimmer.ShimmerFrameLayout
+            android:id="@+id/shimmer_layout"
+            android:layout_width="match_parent"
+            android:layout_height="0dp"
+            app:layout_constraintBottom_toTopOf="@id/pager"
+            app:layout_constraintTop_toBottomOf="@id/recycler_view_cart"
+            app:layout_constraintTop_toTopOf="parent">
```
Shimmer의 가시성도 뷰모델 상태를 통해 데이터바인딩으로 구독할 수 있지 않을까요?

tobae-time replied:
[refactor: 데이터 바인딩을 이용한 로딩 상태 UI 처리](https://github.com/woowacourse/android-shopping-order/pull/104/commits/0aa69752beb4c9e573e5a64df9611ac96ca0cc14)
데이터 바인딩으로 구독하는 것이 더 깔끔한 것 같습니다! 👍

laco-dev commented on app/src/main/res/layout/activity_cart.xml:
```diff
@@ -1,81 +1,94 @@
+                android:background="@color/turquoise"
+                android:padding="10dp"
+                android:gravity="center"
+                android:onClick="@{() -> vm.onChangeScreen()}"
```
뷰모델은 비즈니스 로직을 수행하기 위한 컨트롤러라고 볼 수 있어요.
뷰모델에게 onChangeScreen() 이라고 메시지를 보내면 무엇을 할거라고 기대하시나요?

tobae-time replied:
onChangeScreen() 화면이 바뀌고 나서 실행되는 이벤트 같습니다. 하지만 실제로는 화면 변경을 요청하는 것이므로 requestNavigationToRecommendScreen로 함수명을 변경하였습니다!
[refactor: CartUiEvent.NavigationToRecommendScree, requestNavigationToRecommendScreen로네이밍 변경](https://github.com/woowacourse/android-shopping-order/pull/104/files/0aa69752beb4c9e573e5a64df9611ac96ca0cc14..0c13a13ac64729d175cf5410905d119fcce8c04d)

laco-dev commented on app/src/main/res/layout/activity_cart.xml:
```diff
@@ -1,81 +1,94 @@
-                android:text="@{String.valueOf(vm.uiState.pageState.page)}"
-                android:textColor="@color/gray"
-                android:textSize="22sp" />
+                android:text="전체"
```
stringResource를 사용하고 있다면 일관성 있게 모두 맞춰주세요 :)

tobae-time replied:
[refactor: 전체 선택 문자열 리소스 추가 및 적용](https://github.com/woowacourse/android-shopping-order/pull/104/commits/f786ce1a59eeea046faf3c90de3cd53b0ca03f33)
놓쳤던 부분인 것 같습니다! 앞으로 일관성 있게 모두 맞추도록 하겠습니다! :)

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
-        } else {
-            cartRepository.upsert(productId, decreaseResult.cartQuantity)
+    fun saveHistory(productId: Long) =
+        withState(_uiState.value) { state ->
```
withState는 null을 무시하기 위해 만들어준 상태로 보이는데요
굳이 indent를 늘려가는 것 대비 장점이 무엇일까요?

안그래도 체조 원칙 지켜가면서 열심히 굴렀는데 정면으로 거부하는것이었을까요?
```kotlin
val state = uiState.value ?: return
```

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
+        withState(_uiState.value) { state ->
+            historyRepository.saveHistory(productId) {
+                _uiEvent.postValue(MainUiEvent.NavigateToDetail(productId, state.lastSeenProductId))
+            }
```
그런데 최근 본 상품에 넣는 것은 상품 페이지가 할 일이 아닌가요?
혹시 제가 다른 기능으로 착각했다면 알려주세요.

tobae-time replied:
최근 본 상품을 상품 페이지에서 저장하도록 변경하였습니다!
[refactor: 상품 조회 기록 저장 로직 변경 및 상품 상세 화면 종료 시 기록 동기화](https://github.com/woowacourse/android-shopping-order/pull/104/commits/d5147c310060163277ca5c5e75dd82d09ae5e693)

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
+            val updated = state.decreaseCartQuantity(productId)
+            val cartId = updated.cartId ?: return
+
+            if (updated.hasCartQuantity) {
+                cartRepository.updateQuantity(cartId, updated.cartQuantity) {
+                    it.fold(
+                        onSuccess = { _uiState.value = state.modifyUiState(updated) },
+                        onFailure = { handleError(TAG_DECREASE, it) },
+                    )
+                }
+            } else {
+                cartRepository.deleteCart(cartId) {
+                    it.fold(
+                        onSuccess = {
+                            val result = updated.copy(cartId = null)
+                            _uiState.value = state.modifyUiState(result)
+                        },
+                        onFailure = { handleError(TAG_DECREASE, it) },
+                    )
+                }
```
이 로직은 결국 새로운 상품으로 갈아끼우게 되는데 코드적으로 조금 더 단순하게 표현할 수 있지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
+                }
+
+                else -> {
+                    cartRepository.updateQuantity(cartId, updated.cartQuantity) {
+                        it.fold(
+                            onSuccess = {
+                                _uiState.value = state.modifyUiState(updated)
+                            },
+                            onFailure = { handleError(TAG_INCREASE, it) },
+                        )
+                    }
```
마찬가지로 여기서도 필요한 것 은 새로운 상품을 준비하는 것이고 UI 상태 업데이트는 마지막에 한번만 하면 될 것으로 보여요

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
+        productPage: ProductSinglePage,
+        cartItems: List<ShoppingCart>,
+    ) {
+        historyLoader { result ->
```
유즈케이스를 염두하고 만든 개념 같은데 레포지터리가 여러개의 데이터소스를 사용해도 되잖아요.
왜 HistoryLoader라는 새로운 것을 만들게 된건가요?

만약 다른 사람이 비즈니스 로직을 개발해야 한다면 XXXLoader로 만들어야 할지 레포지터리 사용할지 기준이 없지 않나요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/state/ProductUiState.kt:
```diff
@@ -18,10 +23,26 @@ data class ProductUiState(
+        return copy(productItems = result)
+    }
+
+    fun increaseCartQuantity(productId: Long): ProductState {
```
UiState와 State는 무슨 차이인가요?
이름을 정한 기준이 궁금합니다.

tobae-time replied:
ProductUiState는 상품 목록 화면 전체의 상태를 나타냅니다. 예를 들어, 화면에 보여줄 상품 목록, 사용자가 최근 본 상품 기록, 그리고 상품 데이터를 더 불러올 수 있는지 여부와 같은 전반적인 화면 구성 정보를 포함합니다.

반면, ProductState는 그 목록에 있는 하나의 상품 아이템에 대한 상태를 의미합니다. 이 객체는 개별 상품 정보와 해당 상품이 장바구니에 담긴 상태를 포함하고 있습니다.

따라서 네이밍에 있어 UiState는 화면 단위의 전역 상태를 나타내고, State는 해당 화면을 구성하는 하나의 세부 아이템 상태를 표현하기 위해 사용했습니다!

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/state/ProductState.kt:
```diff
@@ -4,27 +4,26 @@ import woowacourse.shopping.domain.Quantity
 import woowacourse.shopping.domain.product.Product
 
 data class ProductState(
+    val cartId: Long? = null,
```
상품 아이디라는 유니크 정보가 있는데 장바구니 아이디는 어떤 용도로 사용되는지 궁금해요.
지금 구조라면 동일한 상품임에도 불구하고 장바구니 아이디가 달라질 수 있다는 뜻 같아서요.

tobae-time replied:
“동일한 상품임에도 장바구니 아이디가 달라질 수 있다는 뜻 같다"는 말은, 현재 구조가 서버와 너무 강하게 결합되어 있다는 뜻으로 이해하였습니다.
그리고 장바구니 아이디는 API 통신에는 필요하지만, UI에 보여주는 데는 불필요한 정보 같은데,
그런 정보가 UI 상태인 CartState에 들어간 건 불필요한 결합이 아닌가 싶습니다.
혹시 제가 의미를 다르게 해석한 거라면, 더 직관적으로 설명해 주실 수 있을까요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
+
+            when (val cartId = updated.cartId) {
+                null -> {
+                    cartRepository.addCart(Cart(updated.cartQuantity, productId)) {
```
장바구니에 넣는 동작을 서버에서 수행하는데 2초정도 걸린다고 가정해보세요.
유저가 더하기 버튼을 광클하면 어떻게 되나요?

스켈레톤을 위한 로딩 외에도 서버와 통신하기 위한 로딩이라는 상태를 클라이언트 개발자라면 반드시 익숙해 져야 합니다.

laco-dev commented on app/src/main/java/woowacourse/shopping/view/detail/vm/DetailViewModel.kt:
```diff
@@ -84,10 +93,38 @@ class DetailViewModel(
+                        }
+                    },
+                    onFailure = {},
+                )
```
요기서 하고 싶은 것은 장바구니에 담는 협력이라면 카트 레포지터리가 스스로 처리하면 되지 않나요?
이를 위한 모든 로직을 뷰모델에 구현할 필요가 없어보여요

```kotlin
cartRepository.addProduct(...)
```

laco-dev commented on app/src/main/java/woowacourse/shopping/view/detail/vm/DetailViewModel.kt:
```diff
@@ -30,38 +30,47 @@ class DetailViewModel(
-                    ),
-                )
-            }
+        defaultProductRepository.loadProduct(productId) { product ->
```
마찬가지로 상품 레포지터리가 스스로 일을 할 수 있도록 만들어주세요.
가독성이 매우 떨어집니다.

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/vm/CartViewModel.kt:
```diff
@@ -26,37 +26,61 @@ class CartViewModel(
+                result.fold(
+                    onSuccess = { _uiState.value = state.modifyUiState(newState) },
+                    onFailure = {},
+                )
```
UI 상태가 알게모르게 로직을 점점 흡수하는 것 같은데요. 수량을 증가시키는 개념을 UI 상태에서 구현하는게 맞나요?
비즈니스 로직의 결과 새로운 UI 상태로 업데이트가 될 것으로 기대하는데요.
일단 상태를 바꾸고 그 상태 기반으로 비즈니스 로직을 요청하는 것이 순서가 잘못된 것으로 보여서요

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/state/CartUiState.kt:
```diff
@@ -1,39 +1,61 @@
+        val mutableItems = items.toMutableList()
+        mutableItems[targetIndex] = targetItem.modifyChecked(check)
+
+        return copy(items = mutableItems)
```
여기저기 불변 컬렉션으로 잘 처리를 해왔는데 여기서는 가변 컬렉션으로 한 이유가 있나요?

tobae-time replied:
불변 컬렉션을 수정, 변경하는 상황에서 사용하게 되었던 것 같습니다... 해당 부분 불변성을 유지하도록 변경하였습니다!
[refactor: MutableList로 변환했던 리스트를 불변성 유지하도록 수정](https://github.com/woowacourse/android-shopping-order/pull/104/commits/d49e5d4a419f8a93b466e40ef105a6c6184e9f53)

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/state/CartState.kt:
```diff
@@ -0,0 +1,54 @@
+    val price: Price,
+    val quantity: Quantity,
+    val checked: Boolean,
+) {
```
노파심에 남겨봐요.
UI 상태를 사용하다보면 지금처럼 대부분의 로직이 UI 상태에 마구마구 들어가게 됩니다.
그러다보면 도메인 모델은 왜 만듦? 이라는 생각이 어느순간 들게 될텐데요.

우리가 만드는 것은 장바구니라는 도메인 모델의 개념이었는지, 그냥 UI 로직을 짜는 것이었는지 생각해볼만해요.
거의 레벨1에서 했던거는 다 필요없어진거죠.. 😱

이에 대해 페어와 함께 생각을 들려주셨으면 좋겠어요.

tobae-time replied:
페어와 함께 이야기해본 결과, 지금처럼 UI 상태에 로직이 몰리는 구조는 결국 도메인 모델의 역할을 약하게 만들고 있다는 점에 동의하게 되었습니다.
우리가 만들고 있는 것은 단순한 UI가 아니라 ‘장바구니’라는 도메인 개념과 그 안의 비즈니스 규칙이기 때문에, 도메인 모델에 책임을 두고 UI는 그것을 표현하는 역할로 분리하는 게 맞다는 결론을 내렸습니다.
레벨1에서 배운 도메인 설계나 책임 분리 같은 기초가 다시 중요하게 느껴졌고, 앞으로도 UI 상태에 로직을 넣기보단 도메인 모델을 먼저 중심에 놓는 방향으로 가야겠다고 생각했습니다!
해당 부분을 리팩토링한 커밋이 2단계로 함께 묶여 있었던 것 같습니다...!
[refactor: CartState, CartUiState, ShoppingCart 및 Quantity 모델 개선](https://github.com/woowacourse/android-shopping-order/pull/104/commits/5b522b37cf3d6b8844bf7afc3c428a10c3c4f971)

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/adapter/RecommendAdapter.kt:
```diff
@@ -0,0 +1,34 @@
+        holder.bind(items[position])
+    }
+
+    interface Handler : ProductViewHolder.Handler
```
이 인터페이스는 무엇인가요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/OrderCompleteFragment.kt:
```diff
@@ -0,0 +1,54 @@
+
+    private val adapter: RecommendAdapter by lazy {
+        RecommendAdapter(emptyList(), this@OrderCompleteFragment)
+    }
```
이후로는 ListAdapter를 사용해도 괜찮습니다.

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/CartListFragment.kt:
```diff
@@ -0,0 +1,92 @@
+                binding.recyclerViewCart.visibility = View.VISIBLE
+                binding.shimmerLayout.visibility = View.GONE
+            }
+        }
```
이런 내용도 바인딩 어댑터로 옮겨가도 되겠어요

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/CartListFragment.kt:
```diff
@@ -0,0 +1,92 @@
+    private fun setUpBinding(binding: FragmentCartListBinding) {
+        with(binding) {
+            lifecycleOwner = viewLifecycleOwner
+            adapter = cartAdapter
```
어댑터를 바인딩으로 넣어줄 이점이 있나요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -10,44 +10,38 @@ import androidx.appcompat.app.AppCompatActivity
 import woowacourse.shopping.view.cart.vm.CartViewModelFactory
 import woowacourse.shopping.view.core.ext.showToast
-import woowacourse.shopping.view.core.handler.CartQuantityHandler
+import kotlin.io.path.Path
```
Path는 어떤 상황에 사용되는건가요?

laco-dev commented on app/src/main/java/woowacourse/shopping/data/repository/DefaultProductRepository.kt:
```diff
@@ -0,0 +1,53 @@
+                        val productSinglePage = response.toDomain()
+                        callback(Result.success(productSinglePage))
+                    } else {
+                        callback(Result.failure(NullPointerException()))
```
상품이 없다 라는 표현을 전달하기 위해서 nullable 타입도 충분히 고려했을 법 한데요. NPE를 사용한 이유가 있나요?

tobae-time replied:
저번 미션에서 nullable 전파 문제를 겪은 경험이 있어서, 이번에는 nullable보다는 Result를 사용하는 쪽이 유지보수에 더 유리하다고 판단했습니다. 페어와의 논의에서도 실패를 명시적으로 표현하자는 방향에 동의했고, Result.failure()로 처리하게 되었습니다.
다만, NullPointerException은 프로그래밍 오류처럼 보여 오해의 여지가 있을 수 있어, 예외를 따로 정의하여 사용하는 방식이 좋을 것 같습니다. 이 부분은 아직 자신이 없어서 다음 단계를 진행하면서 반영할 수 있도록 해보겠습니다!

laco-dev commented on app/src/main/java/woowacourse/shopping/data/network/response/products/SortX.kt:
```diff
@@ -0,0 +1,14 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class SortX(
```
어디는 Response가 붙어있고 어디는 안붙어있는 이유가 있나요?

tobae-time replied:
강의 자료에 소개된 플러그인을 활용하여 손쉽게 설정을 적용하였습니다. 자동으로 설정을 도와주는 덕분에 빠르게 적용할 수 있었지만, 해당 부분을 믿고 사용하여 따로 확인을 하지 않았습니다. 😅 반성하겠습니다...

laco-dev commented on app/src/main/java/woowacourse/shopping/data/network/response/products/ProductsResponse.kt:
```diff
@@ -0,0 +1,24 @@
+    val size: Int,
+    val sort: SortX,
+    val totalElements: Int,
+    val totalPages: Int,
```
이 모든 데이터가 필요한가요?

tobae-time replied:
모든 응답 데이터를 사용하는 것은 아닙니다. 현재는 content와 last 필드만 사용하고 있습니다. 그래서 나머지 필드를 제거한 상태로 테스트해봤는데, 화면에 데이터가 나타나지 않고 아래와 같은 오류가 발생했습니다!
```
kotlinx.serialization.json.internal.JsonDecodingException: Encountered an unknown key 'totalElements' at offset 4240 at path: $
 Use 'ignoreUnknownKeys = true' in 'Json {}' builder or '@JsonIgnoreUnknownKeys' annotation to ignore unknown keys.
 ```
이를 해결하는 방법은 두 가지가 있습니다.

1. Json { ignoreUnknownKeys = true }로 JSON 파서를 설정하여 모든 알 수 없는 키를 무시하는 방법
2. 각 응답 데이터 클래스에 @JsonIgnoreUnknownKeys 어노테이션을 붙이는 방법

초기에는 문서를 참고해 두 번째 방법인 어노테이션 방식으로 처리했습니다. 그러나 이 방법은 다음과 같은 단점이 있었습니다.

1. 불필요한 필드가 있는 모든 데이터 클래스에 일일이 어노테이션을 붙여야 한다는 점
2. 해당 어노테이션이 실험적 API로, @OptIn(ExperimentalSerializationApi::class)를 함께 사용해야 한다는 점

반면, 첫 번째 방법인 Json { ignoreUnknownKeys = true }는 JSON 빌드를 생성할 때 한 번만 설정하면 됩니다. 이후 해당 빌더를 사용하는 모든 곳에 공통적으로 적용됩니다. 다만, 모든 알 수 없는 키가 무시된다는 특성상 실수로 중요한 키를 누락해도 인지하지 못할 가능성은 있습니다.

하지만 어노테이션 방식이 실험적이라는 점을 고려하여, 저는 경고 없이 안정적으로 동작하는 첫 번째 방법을 선택하여 반영했습니다!

[refactor: Kotlinx Serialization 설정 변경, ignoreUnknownKeys 설정을 Json 인스턴스 생성 시점으로 이동, 불필요한 필드 제거](https://github.com/woowacourse/android-shopping-order/pull/104/commits/72d16fde65e4cd8aa1027d7f35acdfd62087cb01)

laco-dev commented on app/src/main/java/woowacourse/shopping/data/network/response/carts/SortX.kt:
```diff
@@ -0,0 +1,12 @@
+    val sorted: Boolean,
+    @SerialName("unsorted")
+    val unSorted: Boolean,
+)
```
완전히 똑같은 개념이 있는데 다른 패키지에 또 만들 이유가 있나요?

tobae-time replied:
완전히 똑같은 개념을 다른 패키지에 또 만들 이유 없습니다! 하나의 클래스를 재사용하는 방식이 좋을 것 같아, common 패키지로 공통된 클래스를 이동시켰습니다. [refactor: 공통된 Pageable, SortX를 common 패키지로 이동](https://github.com/woowacourse/android-shopping-order/pull/104/commits/3007d59156e43aa63ef42fccd7a44944598aa0b4) 해당 사항을 반영한 커밋입니다!

laco-dev commented on app/src/main/java/woowacourse/shopping/data/network/response/carts/Content.kt:
```diff
@@ -0,0 +1,20 @@
+    val id: Long,
+    val product: Product,
+    val quantity: Int,
+) {
```
Response를 구성하다보면 중첩 클래스를 사용해도 됩니다.
만약 Carts에 다른 API에서도 Content를 사용하면 충돌이 나거든요

tobae-time replied:
현재 response 패키지 하위의 products와 carts에서 각각 동일한 이름의 Content 클래스를 사용하고 있습니다.
다른 패키지에 같은 이름의 Content가 존재할 경우, import 시 충돌이 발생할 수 있습니다.
따라서 중첩 클래스로 변경하여 충돌을 방지하는 것이 좋다는 의미로 이해하여 반영하였습니다.
[refactor: 응답 클래스의 Content 클래스를 중첩 클래스로 변경](https://github.com/woowacourse/android-shopping-order/pull/104/commits/48188989f052ab3723cc61d9ac93b47d8cfcda68)

laco-dev commented on app/src/main/java/woowacourse/shopping/data/network/response/ProductPageEntity.kt:
```diff
@@ -1,4 +1,4 @@
-package woowacourse.shopping.data.network.entity
+package woowacourse.shopping.data.network.response
```
Entity는 왜 response인가요?
무언가 잘못되지 않았나요?

tobae-time replied:
Entity는 데이터베이스에서 사용되는 개체이며 테이블 구조를 매핑하고, Response는 서버로부터 받은 데이터를 담는 객체로 네트워크 통신 결과를 표현합니다.
Entity는 DB와 관련된 곳에, Response는 서버와 관련된 곳에 두는 게 좋을 것 같습니다!
MockServer로부터 응답 받는 객체의 이름이 Entity라는 이름이 어색하다고 느껴진다는 의미로 이해하여 Entity라는 접미사를 Response로 변경해주었습니다!
[refactor: ProductEntity, ProductPageEntity 이름 변경](https://github.com/woowacourse/android-shopping-order/pull/104/commits/a5dbddb6365c063b71e64414e18a0bc4a8a84581)

laco-dev commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -1,25 +1,131 @@
+        quantity: Int,
+        callback: (Result<Unit?>) -> Unit,
+    ) {
+        service.updateCart(cartId, quantity).enqueue(
```
이 구조가 매우 반복되는데 3단계에서 구조가 바뀌어야 하기 때문에 그 때 다시 살펴보겠습니다 :)

laco-dev commented on app/ProductResponse.kt:
```diff
@@ -0,0 +1,12 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ProductResponse(
```
????

tobae-time replied:
[refactor: 중복 ProductResponse 클래스 삭제](https://github.com/woowacourse/android-shopping-order/pull/104/commits/e163f8f345f871363c919684ab6e697fab42f049)
해당 부분이 실수로 중복으로 들어간 것 같습니다... ㅠㅠ 찾아서 제거하였습니다!

laco-dev reviewed (changes requested):
미션 진행 고생 많으셨습니다.

페어와 함께 매우 신나게 이것저것 설계를 고려하고 도입해본 흔적이 보이네요.
다만 제 기준에서 보면 굳이 필요없는 스펙이 많았던 것 같아요.

MVVM 구조에 맞게 잘 구현이 되었지만,
여러가지 개념을 적용하면서 정작 가독성은 완전히 망가졌어요.

나만 보는 이쁜 코드가 아닌 누가 보더라도 이해가 잘되는 코드가 훨씬 가독성이 좋다는 점을 잘 인지하셨으면 좋겠습니다.

chanho0908 committed:
refactor: CartViewHolder 팩토리 메서드 추가

CartViewHolder 생성을 위한 `of` 팩토리 메서드를 추가하여 뷰 바인딩 로직을 캡슐화하고 인스턴스 생성을 단순화합니다.

- `CartViewHolder.kt`: `of` 팩토리 메서드 추가
- `CartAdapter.kt`: `CartViewHolder.of`를 사용하여 ViewHolder 인스턴스 생성

chanho0908 committed:
feat: 추천 상품 UI 구현

- `item_recomment.xml`: 추천 상품 아이템 레이아웃 추가
- `RecommendViewHolder`: 추천 상품 아이템 ViewHolder 구현
- `RecommendAdapter`: 추천 상품 목록 Adapter 구현
- `RecommendDiffUtil`: 추천 상품 목록 DiffUtil 구현

chanho0908 committed:
feat: 상품 목록 화면 UI 이벤트 개선 및 장바구니 이동 로직 추가

chanho0908 committed:
feat: 쉬머 로딩 UI 개선

chanho0908 committed:
feat: 상품 조회 API에 카테고리 필터링 기능 추가

chanho0908 committed:
refactor: CartState, CartUiState, ShoppingCart 및 Quantity 모델 개선

chanho0908 committed:
refactor: 주문 완료 Fragment 이름 변경 및 구현

chanho0908 committed:
refactor: HistoryRepository 및 관련 클래스 수정

chanho0908 committed:
refactor: HistoryRepository 및 관련 클래스 수정

chanho0908 committed:
feat: 주문 완료 화면을 추천 상품 화면으로 변경

chanho0908 committed:
feat: 상품 상세 화면에서 장바구니 이동 시 카테고리 정보 전달

chanho0908 committed:
refactor: MainViewModel의 saveHistory 메서드 파라미터 변경

chanho0908 committed:
refactor: AppContainer 및 BasicAuthentication 코드 정리

chanho0908 committed:
refactor: onClick 핸들러 파라미터에서 category 제거

chanho0908 committed:
refactor: 최근 본 상품 관련 메서드명 변경 및 카테고리 추가

chanho0908 committed:
refactor: CartListFragment 패키지 이동 및 관련 코드 수정

Commit:
feat: 추천 상품 수량 조절 및 UI 상태 관리 기능 추가

Commit:
refactor: CartListFragment의 수량 변경 메서드 파라미터 productId에서 cartId로 변경

chanho0908 committed:
refactor: 불필요한 Log import 제거

- `CartDataSource.kt`: `android.util.Log` import 제거
- `RecommendFragment.kt`: `android.util.Log` import 제거

chanho0908 committed:
refactor: 상품 조회 기록 저장 로직 변경 및 상품 상세 화면 종료 시 기록 동기화

- 상품 조회 기록 저장 시점을 상품 상세 화면 진입 시점으로 변경했습니다.
- 상품 상세 화면 종료 시 `MainActivity`에서 조회 기록을 동기화하도록 변경했습니다.
- `MainActivity`의 `onSelectProduct`, `onClickHistory`에서 `saveHistory` 대신 `handleNavigateDetailEvent`를 호출하도록 수정했습니다.

chanho0908 committed:
refactor: 이벤트 핸들러 ViewModel로 이동

Commit:
fix: 최근 본 상품 로드 시 조회 순서대로 정렬

Commit:
fix: 장바구니에 없는 상품 수량 0으로 설정

Commit:
refactor: ProductEntity, ProductPageEntity 이름 변경

Commit:
refactor: 응답 클래스의 Content 클래스를 중첩 클래스로 변경

Commit:
refactor: 공통된 Pageable, SortX를 common 패키지로 이동

Commit:
refactor: Kotlinx Serialization 설정 변경, `ignoreUnknownKeys` 설정을 `Json` 인스턴스 생성 시점으로 이동, 불필요한 필드 제거

Commit:
refactor: 전체 선택 문자열 리소스 추가 및 적용

Commit:
refactor: item_quantity_changer.xml을 사용하는 곳에서 가시성을 제어하도록 변경

Commit:
refactor: 데이터 바인딩을 이용한 로딩 상태 UI 처리

Commit:
refactor: CartUiEvent.ChangeScreen -> RecommendScreenRequested로 변경 및 관련 코드 수정

Commit:
refactor: CartUiEvent.NavigationToRecommendScree, requestNavigationToRecommendScreen로네이밍 변경

Commit:
fix: CartListFragment에서 RecyclerView가 보이지 않는 문제 해결

Commit:
refactor: MutableList로 변환했던 리스트를불변성 유지하도록 수정

tobae-time reviewed (commented):
안녕하세요, 페로로!
남겨주신 리뷰 중 절반 정도는 코드에 반영했고, 일부는 의견을 남겨두었습니다.
제가 구현 속도가 느린 편이라, 더 늦어지면 다음 단계 진행에 지장이 있을 것 같아 우선 리뷰 재요청 드립니다.

이번에 반영하지 못한 부분과 코멘트를 달지 못한 부분은 다음 단계 진행 중에 꼭 반영할 수 있도록 하겠습니다. 😢

혹시 리뷰 내용을 제가 잘못 이해한 부분이 있다면, 번거로우시겠지만 한 번만 더 코멘트 부탁드립니다!

Commit:
refactor: 중복 ProductResponse 클래스 삭제

laco-dev commented on app/src/main/res/layout/item_recommend.xml:
```diff
@@ -0,0 +1,106 @@
+            app:layout_constraintTop_toBottomOf="@id/image_view_product"
+            tools:text="PET보틀-정사각..." />
+
+        <TextView
+            android:id="@+id/textView_product_price"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="6dp"
+            android:layout_marginEnd="16dp"
+            android:layout_marginBottom="16dp"
+            android:text="@{@string/format_thousand_won(model.item.priceValue)}"
+            android:textColor="@color/black"
+            android:textSize="16sp"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/textView_product_name"
+            tools:text="10,000원" />
```
item_recommend에 대한 UI를 그릴 때 가로 크기를 어디까지 그리는게 의도일까요?
지금의 TextView는 wrap_content이기 때문에 계속해서 늘어날 수가 있어요.

만약 컨테이너 크기를 최대 154까지 하고 싶은 것이라면, ConstraintLayout의 너비가 고정되고 각 자식 View는 match_parent로 그릴 수 있을 것 같아요.

laco-dev commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -192,6 +209,33 @@ class MainViewModel(
         setLoading(false)
     }
 
+    val productEventHandler =
```
구현하다 실수로 가장 아래에 위치된 것 같은데, Handler를 정의하는 것은 뷰모델이 아닌 Activity(UI)가 할 일로 보여요.

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/vm/CartViewModel.kt:
```diff
@@ -126,22 +237,68 @@ class CartViewModel(
         _isLoading.value = isLoading
     }
+
+    val recommendEventHandler =
+        object : RecommendAdapter.Handler {
+            override fun showQuantity(productId: Long) {
+                increaseRecommendProductQuantity(productId)
+            }
+        }
```
실수인줄 알았는데.. 전부 이 패턴으로 작성을 해주셨네요 😓 

이벤트 핸들러를 이곳에, 그리고 뷰모델이 구현한 이유는 무엇일까요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/recommend/RecommendFragment.kt:
```diff
@@ -0,0 +1,56 @@
+        val binding = FragmentRecommendBinding.bind(view)
+        setUpBinding(binding)
+        setUpRecyclerView(binding)
+        viewModel.loadRecommendProduct()
```
화면이 회전하여 매번 호출된다면 비효율적이지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -73,10 +76,11 @@ class CartActivity : AppCompatActivity() {
             is CartListFragment -> {
                 supportFragmentManager.commit {
-                    replace(R.id.fragment_container_view, OrderCompleteFragment())
+                    replace(R.id.fragment_container_view, RecommendFragment())
```
현재 어느 상태에 있고 다음으로 어느 화면으로 이동해야 할지 판단하는 것은 Fragment를 꺼내와서 비교하는 것이 아닌, 뷰모델에서 상태를 판단하여 결정을 해야 하지 않을까요?

laco-dev reviewed (approved):
피드백 반영 고생 많으셨어요.
꼼꼼히 의견 남겨주셨네요.

아직 수정중에 있으시고 다음 단계에서 이어서 리팩토링을 진행하기 때문에 간단한 의견만 몇가지 남겨두었어요.

마지막까지 화이팅입니다 :)
